File: Selah/_contracts/SelahContracts.swift
import Foundation

// MARK: - Shared Selah Contracts

let SelahContractsVersion = "2026-05-25-v1"

enum SelahSafetyTheme: String, Codable, CaseIterable, Equatable, Identifiable {
    case neutral
    case anxiety
    case grief
    case doubt
    case addiction
    case selfHarm
    case abuse
    case trafficking
    case coercion

    var id: String { rawValue }

    var blocksAIDevotionalGeneration: Bool {
        switch self {
        case .selfHarm, .abuse, .trafficking, .coercion:
            return true
        case .neutral, .anxiety, .grief, .doubt, .addiction:
            return false
        }
    }

    var blocksSharing: Bool {
        switch self {
        case .selfHarm, .abuse, .trafficking, .coercion:
            return true
        case .neutral, .anxiety, .grief, .doubt, .addiction:
            return false
        }
    }
}

enum SelahTranslation: String, Codable, CaseIterable, Equatable, Identifiable {
    case kjv = "KJV"
    case esv = "ESV"

    var id: String { rawValue }
}

struct SelahVerseReference: Codable, Hashable, Identifiable {
    let verseId: String
    let translation: SelahTranslation

    var id: String { "\(translation.rawValue):\(verseId)" }
}

struct BereanStudySheetRequest: Codable, Equatable {
    let verseId: String
    let translation: SelahTranslation
    let verseText: String
    let locale: String?
}

struct BereanStudySheetResponse: Codable, Equatable {
    let cacheKey: String
    let verseId: String
    let translation: SelahTranslation
    let layers: BereanStudySheetLayers
    let crossReferences: [String]
    let provenance: BereanStudySheetProvenance
    let generatedAt: Date
    let promptVersion: String

    // The model may receive scripture text as input, but response contracts never
    // include verse text. Clients resolve every verseId from the trusted scripture store.
}

struct BereanStudySheetLayers: Codable, Equatable {
    let text: BereanStudySheetTextLayer
    let context: BereanStudySheetContextLayer
    let interpretation: BereanStudySheetInterpretationLayer
    let application: BereanStudySheetApplicationLayer
}

struct BereanStudySheetTextLayer: Codable, Equatable {
    let observations: [String]
    let keyTerms: [BereanKeyTerm]
    let uncertaintyNotes: [String]
}

struct BereanKeyTerm: Codable, Equatable, Identifiable {
    let id: String
    let term: String
    let note: String
}

struct BereanStudySheetContextLayer: Codable, Equatable {
    let historicalNotes: [String]
    let literaryNotes: [String]
    let canonicalLinks: [String]
}

struct BereanStudySheetInterpretationLayer: Codable, Equatable {
    let summary: String
    let interpretiveOptions: [BereanInterpretiveOption]
    let denominationalPosture: String
    let uncertaintyNotes: [String]
}

struct BereanInterpretiveOption: Codable, Equatable, Identifiable {
    let id: String
    let label: String
    let summary: String
    let confidence: Double
}

struct BereanStudySheetApplicationLayer: Codable, Equatable {
    let prompts: [String]
    let cautions: [String]
    let prayerSeed: String?
}

struct BereanStudySheetProvenance: Codable, Equatable {
    let provider: String
    let model: String
    let runId: String
    let scriptureSource: String
    let scriptureLoadedByClient: Bool
    let factInterpretationSeparated: Bool
}

struct ClassifyVerseThemeRequest: Codable, Equatable {
    let verseId: String
    let translation: SelahTranslation
    let verseText: String
}

struct ClassifyVerseThemeResponse: Codable, Equatable {
    let verseId: String
    let theme: SelahSafetyTheme
    let confidence: Double
    let suggestedActions: [SelahLensActionKind]
    let promptVersion: String
}

enum SelahLensActionKind: String, Codable, CaseIterable, Equatable, Identifiable {
    case understand
    case crossReferences
    case reflect
    case pray
    case addToSession
    case more

    var id: String { rawValue }
}

struct ClassifySafetyRequest: Codable, Equatable {
    let reflectionText: String
    let verseId: String?
    let locale: String?
}

struct ClassifySafetyResponse: Codable, Equatable {
    let theme: SelahSafetyTheme
    let confidence: Double
    let canGenerateDevotional: Bool
    let canShare: Bool
    let supportPayload: SelahSupportPayload?
    let promptVersion: String
}

struct SelahSupportPayload: Codable, Equatable {
    let groundingTitle: String
    let groundingSteps: [String]
    let trustedHumanPrompt: String
    let resourceLinks: [SelahResourceLink]
}

struct SelahResourceLink: Codable, Equatable, Identifiable {
    let id: String
    let title: String
    let url: URL
    let region: String?
}

enum SelahReflectionShareScope: String, Codable, CaseIterable, Equatable, Identifiable {
    case justMe
    case accountabilityPartner
    case namedGroup

    var id: String { rawValue }
}

struct SelahReflectionDocument: Codable, Identifiable, Equatable {
    let id: String
    let ownerUid: String
    let verseId: String?
    let translation: SelahTranslation?
    let body: String
    let safetyTheme: SelahSafetyTheme
    let shareScope: SelahReflectionShareScope
    let sharedWithUid: String?
    let sharedWithGroupId: String?
    let isShareEligible: Bool
    let relationalSignals: SelahRelationalSignals
    let createdAt: Date
    let updatedAt: Date
}

struct SelahRelationalSignals: Codable, Equatable {
    let prayedByGroupCount: Int
    let lastPrayerAt: Date?
}

enum GuidedSelahStep: String, Codable, CaseIterable, Equatable, Identifiable {
    case read
    case listen
    case understand
    case reflect
    case pray
    case apply
    case complete

    var id: String { rawValue }
}

struct GuidedSelahSessionDocument: Codable, Identifiable, Equatable {
    let id: String
    let ownerUid: String
    let verseId: String
    let translation: SelahTranslation
    let currentStep: GuidedSelahStep
    let completedSteps: [GuidedSelahStep]
    let reflectionId: String?
    let cachedStudySheetKey: String?
    let recentThemes: [SelahSafetyTheme]
    let startedAt: Date
    let updatedAt: Date
    let completedAt: Date?
}

struct SelahVerseThemeTagDocument: Codable, Identifiable, Equatable {
    let id: String
    let verseId: String
    let translation: SelahTranslation
    let theme: SelahSafetyTheme
    let confidence: Double
    let promptVersion: String
    let updatedAt: Date
}

struct SelahStudySheetCacheDocument: Codable, Identifiable, Equatable {
    let id: String
    let verseId: String
    let translation: SelahTranslation
    let response: BereanStudySheetResponse
    let promptVersion: String
    let createdAt: Date
    let expiresAt: Date
}

File: Selah/_contracts/selahContracts.ts
export const SELAH_CONTRACTS_VERSION = "2026-05-25-v1" as const;

export type SelahSafetyTheme =
  | "neutral"
  | "anxiety"
  | "grief"
  | "doubt"
  | "addiction"
  | "selfHarm"
  | "abuse"
  | "trafficking"
  | "coercion";

export type SelahTranslation = "KJV" | "ESV";

export type SelahLensActionKind =
  | "understand"
  | "crossReferences"
  | "reflect"
  | "pray"
  | "addToSession"
  | "more";

export interface BereanStudySheetRequest {
  verseId: string;
  translation: SelahTranslation;
  verseText: string;
  locale?: string;
}

export interface BereanStudySheetResponse {
  cacheKey: string;
  verseId: string;
  translation: SelahTranslation;
  layers: BereanStudySheetLayers;
  crossReferences: string[];
  provenance: BereanStudySheetProvenance;
  generatedAt: string;
  promptVersion: string;
  // Scripture text is intentionally absent. Resolve every verseId from the trusted scripture store client-side.
}

export interface BereanStudySheetLayers {
  text: BereanStudySheetTextLayer;
  context: BereanStudySheetContextLayer;
  interpretation: BereanStudySheetInterpretationLayer;
  application: BereanStudySheetApplicationLayer;
}

export interface BereanStudySheetTextLayer {
  observations: string[];
  keyTerms: BereanKeyTerm[];
  uncertaintyNotes: string[];
}

export interface BereanKeyTerm {
  id: string;
  term: string;
  note: string;
}

export interface BereanStudySheetContextLayer {
  historicalNotes: string[];
  literaryNotes: string[];
  canonicalLinks: string[];
}

export interface BereanStudySheetInterpretationLayer {
  summary: string;
  interpretiveOptions: BereanInterpretiveOption[];
  denominationalPosture: string;
  uncertaintyNotes: string[];
}

export interface BereanInterpretiveOption {
  id: string;
  label: string;
  summary: string;
  confidence: number;
}

export interface BereanStudySheetApplicationLayer {
  prompts: string[];
  cautions: string[];
  prayerSeed?: string;
}

export interface BereanStudySheetProvenance {
  provider: string;
  model: string;
  runId: string;
  scriptureSource: string;
  scriptureLoadedByClient: boolean;
  factInterpretationSeparated: boolean;
}

export interface ClassifyVerseThemeRequest {
  verseId: string;
  translation: SelahTranslation;
  verseText: string;
}

export interface ClassifyVerseThemeResponse {
  verseId: string;
  theme: SelahSafetyTheme;
  confidence: number;
  suggestedActions: SelahLensActionKind[];
  promptVersion: string;
}

export interface ClassifySafetyRequest {
  reflectionText: string;
  verseId?: string;
  locale?: string;
}

export interface ClassifySafetyResponse {
  theme: SelahSafetyTheme;
  confidence: number;
  canGenerateDevotional: boolean;
  canShare: boolean;
  supportPayload?: SelahSupportPayload;
  promptVersion: string;
}

export interface SelahSupportPayload {
  groundingTitle: string;
  groundingSteps: string[];
  trustedHumanPrompt: string;
  resourceLinks: SelahResourceLink[];
}

export interface SelahResourceLink {
  id: string;
  title: string;
  url: string;
  region?: string;
}

export type SelahReflectionShareScope = "justMe" | "accountabilityPartner" | "namedGroup";

export interface SelahReflectionDocument {
  id: string;
  ownerUid: string;
  verseId?: string;
  translation?: SelahTranslation;
  body: string;
  safetyTheme: SelahSafetyTheme;
  shareScope: SelahReflectionShareScope;
  sharedWithUid?: string;
  sharedWithGroupId?: string;
  isShareEligible: boolean;
  relationalSignals: SelahRelationalSignals;
  createdAt: FirebaseFirestore.Timestamp;
  updatedAt: FirebaseFirestore.Timestamp;
}

export interface SelahRelationalSignals {
  prayedByGroupCount: number;
  lastPrayerAt?: FirebaseFirestore.Timestamp;
}

export type GuidedSelahStep = "read" | "listen" | "understand" | "reflect" | "pray" | "apply" | "complete";

export interface GuidedSelahSessionDocument {
  id: string;
  ownerUid: string;
  verseId: string;
  translation: SelahTranslation;
  currentStep: GuidedSelahStep;
  completedSteps: GuidedSelahStep[];
  reflectionId?: string;
  cachedStudySheetKey?: string;
  recentThemes: SelahSafetyTheme[];
  startedAt: FirebaseFirestore.Timestamp;
  updatedAt: FirebaseFirestore.Timestamp;
  completedAt?: FirebaseFirestore.Timestamp;
}

export interface SelahVerseThemeTagDocument {
  id: string;
  verseId: string;
  translation: SelahTranslation;
  theme: SelahSafetyTheme;
  confidence: number;
  promptVersion: string;
  updatedAt: FirebaseFirestore.Timestamp;
}

export interface SelahStudySheetCacheDocument {
  id: string;
  verseId: string;
  translation: SelahTranslation;
  response: BereanStudySheetResponse;
  promptVersion: string;
  createdAt: FirebaseFirestore.Timestamp;
  expiresAt: FirebaseFirestore.Timestamp;
}

export function safetyThemeBlocksGeneration(theme: SelahSafetyTheme): boolean {
  return ["selfHarm", "abuse", "trafficking", "coercion"].includes(theme);
}

export function assertNoScriptureTextInStudySheet(response: BereanStudySheetResponse): void {
  const forbiddenKeys = ["verseText", "scriptureText", "textContent", "passageText"];
  const serialized = JSON.stringify(response);
  for (const key of forbiddenKeys) {
    if (serialized.includes(`"${key}"`)) {
      throw new Error(`StudySheetResponse must not contain ${key}; resolve verseId client-side.`);
    }
  }
}
